libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -384,27 +374,135 @@ export class CheckoutService {
         }),
       ]);
 
-      const updatedVersion = version + 1;
-
-      if (!subscription.latest_invoice) {
-        throw new CheckoutError(
-          'latest_invoice does not exist on subscription'
-        );
-      }
-      const latestInvoice = await this.invoiceManager.retrieve(
-        subscription.latest_invoice
-      );
-      try {
-        this.invoiceManager.processPayPalInvoice(latestInvoice);
-      } catch (e) {
-        await this.subscriptionManager.cancel(subscription.id);
-        await this.paypalBillingAgreementManager.cancel(billingAgreementId);
-      }
-
-      await this.postPaySteps(cart, updatedVersion, subscription, uid);
+      this.pollAndUpdatePaymentState(cart.id);
     } catch (error) {
       console.error(error);
       throw error;
     }
   }
+
+  async pollAndUpdatePaymentState(cartId: string) {
+    const startTime = new Date();
+    let msDelay = 500;
+
+    const poll = async () => {
+      try {
+        const cart = await this.cartManager.fetchCartById(cartId);
+        if (cart.state === CartState.FAIL || cart.state === CartState.SUCCESS) {
+          return;
+        }
+        if (!cart.stripeSubscriptionId) {
+          throw new CartSubscriptionNotFoundError(cartId);
+        }
+        if (!cart.uid) {
+          throw new CartError('Cart must have a uid to finalize', {
+            cartId,
+          });
+        }
+        const subscription = await this.subscriptionManager.retrieve(
+          cart.stripeSubscriptionId
+        );
+        if (!subscription) {
+          throw new CartSubscriptionNotFoundError(cartId);
+        }
+
+        const customer = cart.stripeCustomerId
+          ? await this.customerManager.retrieve(cart.stripeCustomerId)
+          : undefined;
+        if (!customer) {
+          throw new CartError('Customer not found for cart', { cartId });
+        }
+
+        // Process PayPal payment
+        if (subscription.collection_method === 'send_invoice') {
```

Comment:
I don't see any polling logic for PayPal payments. Does this logic need to be part of any polling function?

I believe on the call we discussed PayPal having logic for RetryAttempts, however this is only used as part of the `paypal-processor` when processing recurring payments. It's not currently used as part of the Checkout Process.

Reply:
The `poll` local method being run here is largely just a looping method that performs the `pollAndUpdatePaymentState` logic. I'd be happy to rename it, it isn't truly polling at this point.

Reply:
no longer relevant, as the paypal processing logic has been moved into a separate `processPayPalPayment` method